File: Graph/BFS.js
function bfs(G,s)
{
  var color=[];
  var queue=[];
  var d=[0];
  var p=[];
  
  
    var u ;
    
     queue.push(s);
  
                          for(var i=0;i<G.length;i++)
                          {
                            color[i]="white";
                          }
                                      
                                     while( queue.length !== 0 )
                                     {
                                       
                                     
                                      u=queue.shift();
                                     
                                        for( var i=0;i<G.length ;i++)
                                            {
                                                if( G[i].vertex == u)
                                                    {
                                                      var j=i;
                                                     }     
                                            }
                                    
                                      for( var i=0 ;i<G[j].edges.length;i++)
                                        {
                                          for(var k=0 ;k<G.length;k++)
                                          {
                                            if( G[j].edges[i]== G[k].vertex )
                                            {
                                              var woo= k;
                                            }
                                          }
                                          
                                          
                                          if( color[woo] == "white" )
                                            {
                                              color[woo]="gray";
                                              d[woo]=d[j]+1;
                                              p[i]=u; 
                                              queue.push( G[j].edges[i] );
                                            }
                                        }
                                       
                                        color[j]="Black";
                                     }
 return  p;
}

function data(vertex,edges)
{
  this.vertex=vertex;
  this.edges=edges;
}

var G=[
  
 new data("a",["b","d"]),
  new data("b",["a","d","e"]),
  new data("d",["b","g","f"]),
  new data("e",["b","g","i"]),
  new data("f",["d","g","h"]),
  new data("g",["d","e","i","f","h"]),
  new data("h",["f","g","i"]),
  new data("i",["e","g","h"])
  
];

console.log(bfs(G,"a"));

File: Graph/bfs.cpp
#include<iostream>
#include <queue>
using namespace std;
int graph[1000][1000];
int vertices ;
string color[100];
int d[50];
int p[50];

int bfs( int source)
{

  queue<int> myqueue;
  int u ;
  d[0]=0;

  myqueue.push(source);


  for(int i=0;i<vertices;i++)
    {
      color[i]="white";
    }


  while( !myqueue.empty() )
    {
        u =myqueue.front();
            for( int v=0 ;v<vertices ;v++)
                {
                    if( graph[u][v]==1 && color[v]=="white")
                        {
                            color[v]="gray";
                            d[v]=d[u]+1;
                            p[v]=u;
                            myqueue.push(v);

                        }
                }
        color[u]="black";
   }


}

int main()
{
    int row,col,source;
    cout<<"Enter the number of vertices : "<<endl;
    cin>>vertices;
    row=vertices;
    col=row;


        for(int i=0;i<row;i++)
            {
                for(int j=0;j<col;j++)
                    {
                        cin>>graph[i][j];
                    }
            }

    cout<<"  The graph  is :"<<endl;

        for(int i=0;i<row;i++)
            {
                for(int j=0;j<col;j++)
                    {
                        cout<<graph[i][j]<<" ";
                    }
                cout<<endl;
            }

        cout<<"Enter the source vertex for BFS operation :"<<endl;
        cin>>source ;

        bfs(source);

    for(int i=0;i<vertices;i++)
    {
       cout<<" "<<color[i];
    }
    return 0;
}
